### PR TITLE
refactor / simplify reconcileSpec()

### LINKED ISSUE #11262: refactor: simplify reconcileSpec in restore.go
The function [`reconcileSpec`](https://github.com/kubevirt/kubevirt/blob/main/pkg/storage/snapshot/restore.go#L414C27-L414C40) is very complex and have a lot of loops and nested if statements. The goal is to split this in multiple functions and to reduce the if statement nesting complexity

This would be a good issue for people willing to get more familiar with the code for GSoC

### PR DESCRIPTION
**Pull request by coderjojo**
### What this PR does
Before this PR:
reconcileSpec() is difficult to understand due to its complexity and presence of nested statements.

After this PR:
Refactor and split-up reconcileSpec() func present in `pkg/storage/snapshot/restore.go` into multiple functions.

Fixes #11262  
### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: 

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note

```release-note
NONE
```

### CODE REVIEW

File: pkg/storage/snapshot/restore.go [hunk 1]
```diff
@@ -820,6 +798,49 @@ func (t *vmRestoreTarget) doesTargetVMExist() bool {
...
+	}
+
+	// update Running state in case snapshot was on online VM
+	if newVM.Spec.RunStrategy != nil {
+		runStrategyHalted := kubevirtv1.RunStrategyHalted
+		newVM.Spec.RunStrategy = &runStrategyHalted
+	} else if newVM.Spec.Running != nil {
+		running := false
+		newVM.Spec.Running = &running
+	}
```
Comment (alicefr) [inline pkg/storage/snapshot/restore.go:836]: We could use here a switch case instead of a else if

File: pkg/storage/snapshot/restore.go [hunk 2]
```diff
@@ -917,6 +938,27 @@ func (ctrl *VMRestoreController) getDV(namespace, name string) (*v1beta1.DataVol
...
+func getDVTIdx(v kubevirtv1.Volume, snapshotVM *snapshotv1.VirtualMachine) int {
+	templateIndex := -1
+	for i, dvt := range snapshotVM.Spec.DataVolumeTemplates {
+		if v.DataVolume.Name == dvt.Name {
+			templateIndex = i
+			break
+		}
+	}
+	return templateIndex
+}
```
Comment (alicefr) [inline pkg/storage/snapshot/restore.go:960]: We can simplify this a bit:
```suggestion
func getDVTIdx(v kubevirtv1.Volume, snapshotVM *snapshotv1.VirtualMachine) int {
        for i, dvt := range snapshotVM.Spec.DataVolumeTemplates {
                if v.DataVolume.Name == dvt.Name {
                        return i
                }
        }       
        return -1
}
```

File: pkg/storage/snapshot/restore.go [hunk 3]
```diff
@@ -98,6 +98,86 @@ func restoreDVName(vmRestore *snapshotv1.VirtualMachineRestore, name string) str
...
+		pvc, err := t.controller.getPVC(t.vmRestore.Namespace, vr.PersistentVolumeClaimName)
+		if err != nil {
+			return false, err
+		}
+
+		if pvc == nil {
+			return false, fmt.Errorf("pvc %s/%s does not exist and should", t.vmRestore.Namespace, vr.PersistentVolumeClaimName)
+		}
+
+		if nv.DataVolume != nil {
```
[the remark this answers is not in the document]
Comment (coderjojo) [reply pkg/storage/snapshot/restore.go:144]: Yes, make sense.

File: pkg/storage/snapshot/restore.go (hunk 1 above)
Comment (coderjojo) [reply pkg/storage/snapshot/restore.go:836]: Although I thought using switch case won't really simplify it but looking at the rest of code base similar pattern is present, so right to do it.

File: pkg/storage/snapshot/restore.go [hunk 4]
```diff
@@ -98,6 +98,87 @@ func restoreDVName(vmRestore *snapshotv1.VirtualMachineRestore, name string) str
 	return restorePVCName(vmRestore, name)
 }
 
+func updatePVCOwner(vr *snapshotv1.VolumeRestore, pvc *corev1.PersistentVolumeClaim, t *vmRestoreTarget) (bool, error) {
+	if vr.DataVolumeName == nil {
```
Comment (alicefr) [inline pkg/storage/snapshot/restore.go:102]: Can you please flip the condition and return earlier. In this way, we can avoid an additional nesting level

**Comment by alicefr** [comment]
The code looks good, thanks!
/ok-to-test

**Comment by alicefr** [comment]
@coderjojo can you please squash the commit into single one and add `NONE` to the release note?

**Comment by coderjojo** [comment]
> @coderjojo can you please squash the commit into single one and add `NONE` to the release note?

Done. Thanks!

**Comment by alicefr** [comment]
@coderjojo it seems there are some issues with this modification. The functional test related to the restore are failing. You can see them in the `pull-kubevirt-e2e-k8s-1.29-sig-storage`. I would suggest you to reproduce them locally. You can create a local cluster with kubevirtci and foucus on that specific tests by adding an `F` in from of `Context` or `It`

**Comment by avlitman** [comment]
Hi @@coderjojo, thanks for this pr, are you planing on working on this more?

**Pull request closed without merging**